### PR TITLE
fmt::format conversions

### PR DESCRIPTION
**Pull request by neheb**
using std::string's append is inefficient and costs performance. format
is also cleaner.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

**Comment by neheb** [comment]
```
C++ exception with description "invalid format string" thrown in the test body.
```

oh boy...

**Comment by neheb** [comment]
Turns out my little micro optimization doesn't work:

```
config->append_attribute("xsi:schemaLocation") = fmt::format("{1}{2} {1}{2}.xsd", XML_XMLNS, CONFIG_XML_VERSION).c_str();
```

File: src/cds_objects.cc [hunk 1]
```diff
@@ -126,7 +126,7 @@ std::shared_ptr<CdsObject> CdsObject::createObject(unsigned int objectType)
     } else if (IS_CDS_ITEM(objectType)) {
         obj = std::make_shared<CdsItem>();
     } else {
-        throw_std_runtime_error("invalid object type: " + fmt::to_string(objectType));
+        throw_std_runtime_error(fmt::format("invalid object type: {}", objectType));
```
Comment (KarlStraussberger) [inline src/cds_objects.cc:129]: I recommend updating `throw_std_runtime_error(msg)` in `exceptions.h` to include `fmt::format`. Make code a lot clearer and remove another bunch of string concats!

**Review by KarlStraussberger — CHANGES_REQUESTED** [review]

**Comment by KarlStraussberger** [comment]
> 
> 
> Turns out my little micro optimization doesn't work:
> 
> ```
> config->append_attribute("xsi:schemaLocation") = fmt::format("{1}{2} {1}{2}.xsd", XML_XMLNS, CONFIG_XML_VERSION).c_str();
> ```

I'm not sure, but I suppose counting starts at 0. So it has to be `{0}{1}`.

File: src/cds_objects.cc (hunk 1 above)
Comment (neheb) [reply src/cds_objects.cc:129]: That's already the case with this PR.

**Comment by neheb** [comment]
> 
> 
> > Turns out my little micro optimization doesn't work:
> > ```
> > config->append_attribute("xsi:schemaLocation") = fmt::format("{1}{2} {1}{2}.xsd", XML_XMLNS, CONFIG_XML_VERSION).c_str();
> > ```
> 
> I'm not sure, but I suppose counting starts at 0. So it has to be `{0}{1}`.

Weird. In any case, the current version works.

File: src/cds_objects.cc (hunk 1 above)
Comment (KarlStraussberger) [reply src/cds_objects.cc:129]: I was thinking about getting rid of the nesting.

Comment (KarlStraussberger) [reply src/cds_objects.cc:129]: Maybe you can add a second macro  and we reduce it to nested macros.

```
#define throw_runtime_error_with_message(...) throw_std_runtime_error(fmt::format(__VA_ARGS__))
```

**Comment by whyman** [comment]
Nice work @neheb

We can also use `#include <fmt/ostream.h>` to use filesystems built in ostream support with fmt to avoid calling `path.c_str()` everywhere

File: src/config/config_setup.cc [hunk 2]
```diff
@@ -187,21 +186,21 @@ bool ConfigPathSetup::checkAgentPath(std::string& optValue)
...
-            log_error("error in configuration, transcoding profile: could not find transcoding command \"" + optValue + "\"");
+            log_error("Error in configuration, transcoding profile: could not find transcoding command \"{}\"", optValue.c_str());
             return false;
         }
         tmp_path = optValue;
     } else {
         tmp_path = findInPath(optValue);
         if (tmp_path.empty()) {
-            log_error("error in configuration, transcoding profile: could not find transcoding command \"" + optValue + "\" in $PATH");
+            log_error("error in configuration, transcoding profile: could not find transcoding command \"{}\" in $PATH", optValue.c_str());
```
[the remark this answers is not in the document]
Comment (neheb) [reply src/config/config_setup.cc:196]: :). There are a lot more of these.

**Comment by neheb** [comment]
> 
> 
> Nice work @neheb
> 
> We can also use `#include <fmt/ostream.h>` to use filesystems built in ostream support with fmt to avoid calling `path.c_str()` everywhere

Hmm? Isn't the purpose of that to keep size down?

File: src/cds_objects.cc (hunk 1 above)
Comment (whyman) [reply src/cds_objects.cc:129]: fmt happily handles the "no-args" case, so just updating the existing macro should work

**Comment by neheb** [comment]
Alright. Seems this PR also gets the size down:

```
4481928
4346536
```

**Comment by neheb** [comment]
Oh I see. The macro is wrong...

File: src/exceptions.h [hunk 3]
```diff
@@ -43,8 +43,7 @@
 #define PRETTY_FUNCTION __func__
 #endif
 
-#define throw_std_runtime_error(msg) throw std::runtime_error(std::string(msg) \
-    + " file:" + std::string(__FILE__) + " line:" + fmt::to_string(__LINE__) + " function:" + std::string(PRETTY_FUNCTION))
+#define throw_std_runtime_error(...) throw std::runtime_error(fmt::format("{} file: {} line: {} functions: {}", fmt::format(__VA_ARGS__), __FILE__, __LINE__, PRETTY_FUNCTION))
```
Comment (KarlStraussberger) [inline src/exceptions.h:47]: Well, it's `functions` instead of `function`. What about turning things around: "ERROR in file {}, line {}, function {}: {}"

**Review by KarlStraussberger — APPROVED** [review]

**Comment by KarlStraussberger** [comment]
@neheb  Really nice! No more formatting issues when texting exections!

File: src/exceptions.h (hunk 3 above)
Comment (whyman) [reply src/exceptions.h:47]: Perhaps a format similar to the debug messages, nice and compact:

`"[{}:{}] {} Error: {}", __FILE__, __LINE__, PRETTY_FUNCTION, fmt::format(__VA_ARGS__))`

Comment (neheb) [reply src/exceptions.h:47]: Switched to this format string.

**Pull request merged**